### PR TITLE
feat(loss): add --loss-aggregation pg_loss modes

### PR DESCRIPTION
**Pull request by EazyReal**
Port of THUDM/slime#2090 for miles.

Summary:
- add `--loss-aggregation {sample_mean,prompt_mean,token_mean,constant}` for pg_loss
- keep `sample_mean` as the default and reconcile legacy `--calculate-per-token-loss` with `token_mean`
- sync the current slime `prompt_mean` behavior: rollout conversion emits `prompt_mask_sums`, the reducer scales by `n_samples_per_prompt`, and the final scalar is the direct mean over prompt groups
- validate `prompt_mean` configs with `global_batch_size % n_samples_per_prompt == 0`, so each prompt group stays whole within a train step
- add constant-divisor aggregation for Dr.GRPO while keeping non-pg metrics on the default reducer
- include docs and focused CPU coverage for reducer math, argument validation, and prompt_mean rollout metadata

Validation:
- `uv run --with pytest --with torch --with numpy --with httpx --with pyyaml --with ray --with huggingface_hub --with transformers --with pydantic pytest --confcutdir=tests/fast/backends/training_utils tests/fast/backends/training_utils/test_loss_aggregation.py -q` -> 27 passed
- `uv run --with ruff ruff check miles/backends/training_utils/cp_utils.py miles/backends/training_utils/data.py miles/backends/training_utils/loss.py miles/backends/training_utils/loss_hub/losses.py miles/ray/rollout/train_data_conversion.py miles/utils/arguments.py miles/backends/megatron_utils/model.py miles/backends/experimental/fsdp_utils/actor.py tests/fast/backends/training_utils/test_loss_aggregation.py tests/fast/backends/training_utils/loss/test_loss_snapshot.py` -> passed
- `uv run --with black black --check miles/backends/training_utils/cp_utils.py miles/backends/training_utils/data.py miles/backends/training_utils/loss.py miles/backends/training_utils/loss_hub/losses.py miles/ray/rollout/train_data_conversion.py miles/utils/arguments.py miles/backends/megatron_utils/model.py miles/backends/experimental/fsdp_utils/actor.py tests/fast/backends/training_utils/test_loss_aggregation.py tests/fast/backends/training_utils/loss/test_loss_snapshot.py` -> passed
- `git diff --check upstream/main` -> passed

### CODE REVIEW

File: miles/backends/training_utils/cp_utils.py [hunk 1]
```diff
@@ -93,22 +94,40 @@ def get_sum_of_sample_mean(
...
+    batches would otherwise get a partial denominator on each side.
+
+    ``sample_denoms`` is a per-sample scalar tensor list and is left untouched
+    by CP chunking — only the per-token ``loss_masks`` are sliced into the
+    local zigzag layout, so CP correctness is preserved for every mode.
     """
-    Calculate correct sample mean for CP
-    """
+    if sample_denoms is None:
+        sample_denoms = [loss_mask.sum() for loss_mask in loss_masks]
```
[the remark this answers is not in the document]
Comment (EazyReal) [reply miles/backends/training_utils/cp_utils.py:109]: Pushing back here. `sample_denoms` are already guaranteed on the correct device at reduction time: in production they are built in `data.py:get_rollout_data` via `torch.tensor(d, dtype=torch.float32, device=torch.cuda.current_device())` — the same call used for `loss_masks` — before any reducer runs; the `constant` path builds them with `device=loss_mask.device`; and the `sample_denoms=None` default uses `loss_mask.sum()` (already on-device). So no H2D copy happens in the hot loop, and adding a per-sample `torch.as_tensor(..., device=...)` here would only mask a genuine device bug rather than surface it. On the length check: the reducer's `zip(..., strict=True)` already raises `ValueError` on any length mismatch with zero extra code, so an explicit `len()` guard is redundant.

File: miles/ray/rollout/train_data_conversion.py [hunk 2]
```diff
@@ -55,6 +55,26 @@ def convert_samples_to_train_data(
...
+    # contributions then sum to one token-weighted mean per prompt group
+    # (averaged over groups by the outer /global_batch_size in loss.py).
+    #
+    # sample_mean (None denom), token_mean, and constant are not data-dependent
+    # here and are resolved by the reducer / args; only prompt_mean needs this.
+    if getattr(args, "loss_aggregation", "sample_mean") == "prompt_mean":
+        group_mask_totals: dict[int, int] = {}
+        for sample, loss_mask in zip(samples, loss_masks, strict=True):
+            group_mask_totals[sample.group_index] = group_mask_totals.get(sample.group_index, 0) + sum(loss_mask)
+        train_data["sample_denoms"] = [group_mask_totals[sample.group_index] for sample in samples]
```
[the remark this answers is not in the document]
Comment (EazyReal) [reply miles/ray/rollout/train_data_conversion.py]: Good catch — accepted. `Sample.group_index` is typed `int | None`, so a missing value would silently merge unrelated prompts under `group_mask_totals[None]` and corrupt the denominator. Added a fail-loud `raise ValueError("--loss-aggregation prompt_mean requires every Sample.group_index to be set.")` inside the prompt_mean loop (used `raise` rather than `assert` so it survives `python -O`), with a load-bearing test. The length-mismatch part is already covered by the loop's `zip(..., strict=True)`, so I did not add a separate `len()` check.

**Comment by EazyReal** [comment]
Hi @yueming-yuan, resurfacing this for visibility since the source slime PR has moved forward.

I synced the Miles port with the current THUDM/slime#2090 behavior: `prompt_mean` now uses `prompt_mask_sums`, scales the reducer by `n_samples_per_prompt`, and validates that `global_batch_size` is a multiple of `n_samples_per_prompt`, so the final scalar is the direct mean over prompt groups. The branch is rebased as one commit and the focused pytest/ruff/black checks are listed in the PR body.

Would appreciate your review when you have a chance.

**Comment by EazyReal** [comment]
Closing in favor of a fresh PR from `upstream-pr/loss-aggregation-modes-v2` for visibility. The new branch carries the same cleaned, slime-synced commit.
